ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1278,13 +1282,100 @@ private class FirstGetChunk extends GetChunk {
     private BlobProperties serverBlobProperties;
     private short lifeVersion;
 
+    private AmbryCache blobMetadataCache;
+    private AmbryCacheEntryStatus blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheNotLookedUp;
+
     /**
      * Construct a FirstGetChunk and initialize it with the {@link BlobId} of the overall operation.
+     * @param blobMetadataCache A cache to save blob metadata for composite blobs
      */
-    FirstGetChunk() {
+    FirstGetChunk(AmbryCache blobMetadataCache) {
       super(-1, new CompositeBlobInfo.ChunkMetadata(blobId, 0L, -1L));
+      this.blobMetadataCache = blobMetadataCache;
+      if (shouldLookupMetadataCache()) {
+        BlobMetadata blobMetadata = (BlobMetadata) blobMetadataCache.getObject(blobId.toString());
+        if (blobMetadata != null) {
+          blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheHit;
+          blobType = BlobType.MetadataBlob;
+          compositeBlobInfo = blobMetadata.getCompositeBlobInfo();
+          blobInfo = blobMetadata.getBlobInfo();
+          lifeVersion = blobInfo.getLifeVersion();
+          serverBlobProperties = blobInfo.getBlobProperties();
+          chunkIndexToBuf = new ConcurrentHashMap<>();
+          chunkIndexToBufWaitingForRelease = new ConcurrentHashMap<>();
+          successfullyDeserialized = true;
+        } else {
+          blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheMiss;
+        }
+      }
+    }
+
+    /**
+     * Decides if we should look up the metadata cache during a GET request
+     * @return True if and only if
+     *  - cache exists
+     *  - it is a RANGE request
+     *  - it is not a raw mode request
+     *  - the blobID is for a metadata blob
+     *  - the cache not yet been consulted
+     *  False, otherwise.
+     */
+    boolean shouldLookupMetadataCache() {
+      boolean result = (blobMetadataCache != null &&
+          options.getBlobOptions.getRange() != null &&
```

Comment:
I think we can remove the `options.getBlobOptions.getRange() != null`. Even if we are trying to get the whole blob, we can still leverage cache.

Reply:
That seems to be the case. But we might have multiple clients issuing get requests.

Reply:
Sorry, my reply above was for shouldSaveMetadata. But for shouldLookupMetadataCache, in reply to your comment, we should not look up cache on each GET request. We discussed this in the design doc and Ashish suggested to lookup the cache conditionally for eg. only on range requests. This makes sense since vast majority of blobs are small, most requests are GET for small blobs and will not see much benefits from caching. Looking up the cache on each GET just adds more latency. I have some perf numbers to back it up.

Reply:
> That seems to be the case. But we might have multiple clients issuing get requests.

Just saw this. I think we both commented at the same time. Yes, you are right, we can get multiple GET requests for the same blob from multiple clients. I will change the code to leverage that but we need to set some threshold for blob size below which we won't cache.

Reply:
I thought more about this. If we save metadata even for every GET request on a blob (subject to some min size), then we may want to look it up for subsequent GETs on the same blob. The problem is that we don't know what the size of a blob is just by looking at the blobID. So then we need to lookup the cache for each GET which just adds extra latency for those blobs whose metadata wasn't cached. I discussed this issue in the design doc and adding latency to GETs that will mostly encounter a cache miss wasn't acceptable.